File: mirai-core/src/main/java/net/mamoe/mirai/utils/setting/MiraiSettingMapSection.java
package net.mamoe.mirai.utils.setting;


import org.ini4j.Profile;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;


public class MiraiSettingMapSection extends ConcurrentHashMap<String, Object> implements MiraiSettingSection {

    @SuppressWarnings("unchecked")
    public <T> T get(String key, T defaultValue) {
        if (key == null || key.isEmpty()){
            return defaultValue;
        }
        if (super.containsKey(key)){
            return (T) super.get(key);
        }
        return defaultValue;
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String key) {
        if (key == null || key.isEmpty()) {
            return null;
        }
        if (super.containsKey(key)) {
            return (T) super.get(key);
        }
        return null;
    }

    public void set(String key, Object value){
        this.put(key,value);
    }

    public void remove(String key){
        super.remove(key);
    }

    public int getInt(String key) {
        return this.getInt(key, 0);
    }

    public int getInt(String key, int defaultValue) {
        return Integer.parseInt(String.valueOf(this.get(key, defaultValue)));
    }

    public double getDouble(String key) {
        return this.getDouble(key, 0D);
    }

    public double getDouble(String key, double defaultValue) {
        return Double.parseDouble(String.valueOf(this.get(key, defaultValue)));
    }

    public float getFloat(String key) {
        return this.getFloat(key, 0F);
    }

    public float getFloat(String key, float defaultValue) {
        return Float.parseFloat(String.valueOf(this.get(key, defaultValue)));
    }

    public String getString(String key) {
        return this.getString(key, "");
    }

    public String getString(String key, String defaultValue) {
        return String.valueOf(this.get(key, defaultValue));
    }

    public Object getObject(String key) {
        return this.get(key);
    }

    @SuppressWarnings("unchecked")
    public <T> List<T> asList(){
        return this.values().stream().map(a -> (T)(a)).collect(Collectors.toList());
    }

    @Override
    public synchronized void saveAsSection(Profile.Section section) {
        section.clear();
        this.forEach(section::put);
    }

    @Override
    public void close() throws IOException {

    }

}

File: mirai-core/src/jvmMain/kotlin/net/mamoe/mirai/utils/setting/MiraiSettingMapSection.kt
package net.mamoe.mirai.utils.setting


import org.ini4j.Profile

import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors
import kotlin.streams.toList


/**
 * @author NaturalHG
 */
class MiraiSettingMapSection : ConcurrentHashMap<String, Any>(), MiraiSettingSection {

    operator fun <T> get(key: String?, defaultValue: T): T {
        if (key == null || key.isEmpty()) {
            return defaultValue
        }
        return if (super.containsKey(key)) {
            super.get(key) as T
        } else defaultValue
    }


    operator fun set(key: String, value: Any) {
        this[key] = value
    }

    override fun remove(key: String) {
        super.remove(key)
    }

    fun getInt(key: String): Int {
        return this.getInt(key, 0)
    }

    fun getInt(key: String, defaultValue: Int): Int {
        return Integer.parseInt(this[key, defaultValue].toString())
    }

    fun getDouble(key: String): Double {
        return this.getDouble(key, 0.0)
    }

    fun getDouble(key: String, defaultValue: Double): Double {
        return java.lang.Double.parseDouble(this[key, defaultValue].toString())
    }

    fun getFloat(key: String): Float {
        return this.getFloat(key, 0f)
    }

    fun getFloat(key: String, defaultValue: Float): Float {
        return java.lang.Float.parseFloat(this[key, defaultValue].toString())
    }

    fun getString(key: String): String {
        return this.getString(key, "")
    }

    fun getString(key: String, defaultValue: String): String {
        return this[key, defaultValue].toString()
    }

    fun getObject(key: String): Any? {
        return this[key]
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> asList(): List<T> {
        return this.values.stream().map { a -> a as T }.toList()
    }

    @Synchronized
    override fun saveAsSection(section: Profile.Section) {
        section.clear()
        this.forEach{ key, value -> section.put(key, value) }
    }

    @Throws(IOException::class)
    override fun close() {

    }

}

